### PR TITLE
Object flattener logic

### PR DESCRIPTION
**Pull request by williamluang**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,33 @@
 Dial-Once
 
-Flattener module
\ No newline at end of file
+There are no environment to set up, this module is ready to use.
```
Comment (mrister) [inline README.md]: there **is** no environment...

File: package.json [hunk 2]
```diff
@@ -2,15 +2,20 @@
...
     "node",
     "flattener",
     "camel"
   ],
   "author": "Dial-Once",
-  "license": "ISC"
+  "license": "ISC",
+  "devDependencies": {
+    "eslint": "^4.4.1",
+    "flatten-obj": "^3.1.1",
```
Comment (mrister) [inline package.json]: this should not be a dev dependency since you use it in src/index.js, but a dependecy

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
+const flattener = require('flatten-obj')();
+
+// see http://stackoverflow.com/questions/19098797/fastest-way-to-flatten-un-flatten-nested-json-objects
+// http://jsfiddle.net/WSzec/14/
+/* eslint-disable no-use-before-define, no-param-reassign */
+function loop(cur, prop, result, exclusions) {
+  let isEmpty = true;
+
+  for (const p in cur) {
+    if (Object.prototype.hasOwnProperty.call(cur, p)) {
```
Comment (mrister) [inline src/index.js]: what is someone would like to use the module but using properties that are down the prototype chain ?

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    result[prop] = [];
+  } else if (isEmpty && !Array.isArray(cur)) {
+    result[prop] = {};
+  }
+}
+
+function recurse(cur, prop, result, exclusions) {
+  if (exclusions.includes(prop)) return;
+
+  if (Object(cur) !== cur) {
```
Comment (mrister) [inline src/index.js]: can you explain a bit what is the purpose if this check and why you use Object?

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    if (cur !== undefined) {
+      result[prop] = cur;
+    }
+  } else {
+    loop(cur, prop, result, exclusions);
+  }
+}
+
+module.exports = function (data, exclusions) {
```
Comment (mrister) [inline src/index.js]: arrow function ?

Comment (mrister) [inline src/index.js]: do you need to export this at all since camelCase is the only exported function being used ?

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  recurse(data, '', result, exclusions);
+  return result;
+};
+
+module.exports.camelCase = (obj) => {
+  const result = {};
+  // this one converts an obj to { key.innerkey: value } format
+  const flattenedObj = flattener(obj);
+  // replacing next char after dot (.) with the same but in UpperCase and removing dots
+  for (const key of Object.keys(flattenedObj)) {
```
Comment (mrister) [inline src/index.js]: use Object.keys(flattenedObj)  before the for loop to skip looking up keys on every itereation

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    const value = flattenedObj[key];
+    let camelCaseKey = '';
+    const regex = /\./gi;
```
Comment (mrister) [inline src/index.js]: suggestion HAS_DOT_REGEX

File: test/index-spec.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const flattener = require('../src/index.js');
+const assert = require('assert');
+
+describe('flattener camelCase', () => {
+  it('should flatten object to camel-case style', () => {
```
Comment (mrister) [inline test/index-spec.js:5]: would be nice to have more test cases, objects with prototypes, numbers, array, null and any other edge case you could think of.

**Review by mrister — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 6 above)
Comment (Spring3) [reply src/index.js]: we checked that and it only looks there once

Comment (Spring3) [reply src/index.js]: `for of` looks once while plain `for` loop calculates the source each iteration

File: test/index-spec.js (hunk 8 above)
Comment (Spring3) [reply test/index-spec.js:5]: won't work for array. The purpose was to flatten plain objects. Even the external library that actually does flattening was picked coming out of the idea that only objects will be flattened

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+module.exports = function (data, exclusions) {
+  const result = {};
+
+  exclusions = exclusions || [];
+
+  recurse(data, '', result, exclusions);
+  return result;
+};
+
+module.exports.camelCase = (obj) => {
```
Comment (Spring3) [inline src/index.js]: @williamluang sorry for not mentioning this during our call. You only need this function. Other functions should not be included into the library.

**Review by Spring3 — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 9 above)
Comment (Spring3) [reply src/index.js]: Also maybe you could think of a way to create a small layer of abstraction and make it a place for configuration/initialization. For example to support not only env variables, but also plain settings object that will affect the way the library works. Not everyone configures the app using env vars prior to settings. And @mrister also mentioned this when he asked to have some ability to pass for example HASHING_KEY as a config property

Comment (Spring3) [reply src/index.js]: In case you need some help, do not hesitate to ask

Comment (williamluang) [reply src/index.js]: Ok thank you, I need to document myself a bit but i'll ask you if there are things that I have trouble understanding !

**Review by ofkindness — APPROVED** [review]

File: src/index.js (hunk 6 above)
Comment (mrister) [reply src/index.js]: nice to know!

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  const result = {};
+  // this one converts an obj to { key.innerkey: value } format
+  const flattenedObj = flattener(obj);
+  // replacing next char after dot (.) with the same but in UpperCase and removing dots
+  for (const key of Object.keys(flattenedObj)) {
+    const value = flattenedObj[key];
+    let camelCaseKey = '';
+    const HAD_DOT_REGEX = /\./gi;
```
Comment (mrister) [inline src/index.js]: suggestion: HAS_DOT_REGEX

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    // if key contains dot
```
Comment (mrister) [inline src/index.js]: quite obvious from code, you can remove such comments

File: test/index-spec.js (hunk 8 above)
Comment (mrister) [reply test/index-spec.js:5]: Should be as noted in readme. but other cases would be nice

**Review by mrister — CHANGES_REQUESTED** [review]

**Review by Spring3 — APPROVED** [review]
please add some info about supported object types and we are good to go

**Review by ykhatal — APPROVED** [review]

File: README.md [hunk 12]
```diff
@@ -1,3 +1,35 @@
...
-Flattener module
\ No newline at end of file
+There is no environment to set up, this module is ready to use.
+It flattens object literal into camel case format.
+
+A JavaScript object literal is a comma-separated list of name-value pairs wrapped in curly braces (sources: http://www.dyn-web.com/tutorials/object-literal/)
+
+There is only one function -> camelCase();
+
+Exammples:
```
Comment (jkernech) [inline README.md]: ### Typo 
> Examples

File: README.md [hunk 13]
```diff
@@ -1,3 +1,35 @@
 Dial-Once
```
Comment (jkernech) [inline README.md]: Not sure the meaning of this.. Why not use the module name ?

File: sonar-project.properties [hunk 14]
```diff
@@ -0,0 +1,11 @@
+sonar.projectKey=flat-camel
+sonar.projectName=flat-camel
+sonar.projectVersion=0.0.0
```
Comment (jkernech) [inline sonar-project.properties]: This can be removed

File: src/index.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+const flattener = require('flatten-obj')();
+
+// see http://stackoverflow.com/questions/19098797/fastest-way-to-flatten-un-flatten-nested-json-objects
+// http://jsfiddle.net/WSzec/14/
+/* eslint-disable no-use-before-define, no-param-reassign */
+
+module.exports.camelCase = (obj) => {
```
Comment (jkernech) [inline src/index.js]: ### Suggestion
> toCamelCase

**Review by jkernech — CHANGES_REQUESTED** [review]
The README needs to be updated to be more consistent, to be more concise the description of the module as well as the documentation needs to be reworked

File: sonar-project.properties [hunk 16]
```diff
@@ -0,0 +1,10 @@
+sonar.projectKey=flat-camel
```
Comment (mrister) [inline sonar-project.properties]: node-flat-camel ?

File: src/index.js [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    for (let i = 0; i < object.length; i++) {
+      result[i] = module.exports.toCamelCase(object[i]);
+    }
+  } else {
+    // this one converts an object to { key.innerkey: value } format
+    const flattenedObj = flattener(object);
+    // replacing next char after dot (.) with the same but in UpperCase and removing dots
+    for (const key of Object.keys(flattenedObj)) {
+      let camelCaseKey = key;
+      const CHAR_NEXT_TO_DOT = /\../gi;
```
Comment (mrister) [inline src/index.js]: there is no need to redefine the const for each for loop it can be moved up

**Review by mrister — APPROVED** [review]

**Review by ykhatal — APPROVED** [review]

**Review by ofkindness — APPROVED** [review]

**Review by jkernech — APPROVED** [review]
Aside from @mrister comment and the README.md format, all good
